src/rust/inetstack/protocols/tcp/peer.rs

```diff
@@ -507,62 +567,82 @@ impl Inner {
         let nonce: u32 = rng.gen();
         Self {
             isn_generator: IsnGenerator::new(nonce),
-            ephemeral_ports,
-            sockets: HashMap::new(),
-            passive: HashMap::new(),
-            connecting: HashMap::new(),
-            established: HashMap::new(),
-            rt,
-            scheduler,
-            clock,
-            local_link_addr,
-            local_ipv4_addr,
-            tcp_config,
-            arp,
+            ephemeral_ports: ephemeral_ports,
+            rt: rt,
+            scheduler: scheduler,
+            qtable: qtable.clone(),
+            addresses: HashMap::<SocketId, QDesc>::new(),
+            clock: clock,
+            local_link_addr: local_link_addr,
+            local_ipv4_addr: local_ipv4_addr,
+            tcp_config: tcp_config,
+            arp: arp,
             rng: Rc::new(RefCell::new(rng)),
-            dead_socket_tx,
+            dead_socket_tx: dead_socket_tx,
         }
     }
 
-    fn receive(&mut self, ip_hdr: &Ipv4Header, buf: DemiBuffer) -> Result<(), Fail> {
+    fn receive(&self, ip_hdr: &Ipv4Header, buf: DemiBuffer) -> Result<(), Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::receive");
```

Comment:
Could we add changes that concern performance observation in a different commit?